[dynamo] np.sort(complex) is not implemented

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107689
* #107746
* #107711
* __->__ #107710
* #107688
* #107687

This issue was discovered once we were able to trace without breaking
in https://github.com/pytorch/pytorch/pull/107689. Same for the next
one.

cc @mruberry @rgommers @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

lezcano committed:
[dynamo] np.sort(complex) is not implemented

This issue was discovered once we were able to trace without breaking
in https://github.com/pytorch/pytorch/pull/107689. Same for the next
one.

[ghstack-poisoned]

vadimkantorov commented:
Funnily, related: https://github.com/pytorch/pytorch/issues/58034#issuecomment-1687279902

ev-br commented on torch/_numpy/_funcs_impl.py:
```diff
@@ -1387,13 +1387,17 @@ def _sort_helper(tensor, axis, kind, order):
 
 def argsort(a: ArrayLike, axis=-1, kind=None, order: NotImplementedType = None):
+    if a.dtype.is_complex:
+        return NotImplemented
```
Not: Can move to _sort_helper to remove duplication.

lezcano replied:
but would need to return somehow from the main function, so I didn't really know how to do it uniformly. I figured we could always just raise a `NotImplementedError` and then catch it in the normaliser and then return...

lezcano replied:
Given that, a the moment, we always raise, we should probably go with the normaliser approach.

The point as to why doing the return rather than the exception is that dynamo already knows how to treat a function that returns `NotImplemented`, falling back gracefully to eager.

ev-br reviewed (commented):
Optional: can also add same check to searchsorted.

I suppose we should either always return NotImplemented or always raise NotImplementedError? (Here and in normalize_not_implemented). As of this PR, using the order kwarg raises, but sorting a complex array returns.

lezcano committed:
Update on "[dynamo] np.sort(complex) is not implemented"


This issue was discovered once we were able to trace without breaking
in https://github.com/pytorch/pytorch/pull/107689. Same for the next
one.

cc mruberry rgommers voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

lezcano commented:
Good point. Added searchsorted.

ezyang commented:
@pytorchbot merge

lezcano committed:
Update on "[dynamo] np.sort(complex) is not implemented"


This issue was discovered once we were able to trace without breaking
in https://github.com/pytorch/pytorch/pull/107689. Same for the next
one.

cc mruberry rgommers voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

lezcano committed:
Update on "[dynamo] np.sort(complex) is not implemented"


This issue was discovered once we were able to trace without breaking
in https://github.com/pytorch/pytorch/pull/107689. Same for the next
one.

cc mruberry rgommers voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]
